### PR TITLE
Shift Most of Wasm AOT test build to helix

### PR DESCRIPTION
**Pull request by steveisok**
Since AOT'ing each test suite takes between 3-9 min, we need to shift the burden over to helix.

## This is done by:
1. building the test assemblies on the build machine
   - the wasm part of the build is not executed on the build machine,
      because it has the AOT build part
2. Zip up the test assembly+friends, *and* any bits required to run the wasm
    app build for that on helix (eg. emsdk, wasm app targets, cross compiler etc)
3. Send all this to helix, and use a custom `aot-build.proj`
     - which recreates all the build inputs for the `WasmBuildApp` target
        using the paths for the assets on helix
     - then we can run `WasmBuildApp` for the build, resulting in a wasm app
       bundle.
4. Run the tests!

- We already have the bits required for building wasm apps on helix, supported
for `Wasm.Build.Tests`, which we can use here too.

## Trimming:

- Since, AOT can be so expensive, we use `EnableAggressiveTrimming=true`(EAT), but
   that means that we could have issues due to trimming.
- And it can sometimes be unclear whether the build/test failures are due to trimming
     or AOT.

- Because these builds+test runs are different from other builds, owing to the
   "build partially on helix" step, a normal EAT build would not be the same as 
- to help with testing this, we add two lanes to `runtime-staging`:
   - `*_Mono_AOT`: builds AOT+EAT on helix
   - `*_Mono_EAT`: builds EAT, on helix
      - this is required because we want to run almost the same kinda
         build: 1. build test assembly; *2. send to helix; 3. build wasm app;* 4. run tests

- This should effectively mean that we can see which errors might be due to EAT, and
   which are clearly because of EAT+AOT.

### CODE REVIEW

File: src/mono/wasm/build/aot-build.proj [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<Project DefaultTargets="BundleTestWasmApp">
```
Comment (radical) [inline src/mono/wasm/build/aot-build.proj:1]: Drop `Directory.Build.{props,targets}` like https://github.com/dotnet/runtime/blob/main/src/tests/BuildWasmApps/Wasm.Build.Tests/WasmBuildAppTest.cs#L620-L660 next to this aot-build.proj on helix. And aot-build.proj doesn't need to setup all the paths, since it will get set up by `WasmApp.LocalBuild*`.

**Review by radical — COMMENTED** [review]
Now that Wasm.Build.Tests are merged. This can use the bits enabled by that.

- For the build tests, we bundle up all the bits required for building wasm apps on helix. That can be utilized here too (https://github.com/dotnet/runtime/blob/main/src/libraries/sendtohelixhelp.proj#L251-L257)
   - That puts all of that under `${HELIX_CORRELATION_PAYLOAD}/build`

- Instead of `aot-build.proj` setting up paths itself, you can drop `Directory.Build.{props, targets}` like https://github.com/dotnet/runtime/blob/main/src/tests/BuildWasmApps/Wasm.Build.Tests/WasmBuildAppTest.cs#L620-L660 next to `aot-build.proj`.
  - That will set up all the paths for you (as it uses `WasmApp.LocalBuild*`, which is meant exactly for this use case)
  - and `aot-build.proj` can just setup anything extra that it needs

- The final thing needed would be to pass `/p:WasmBuildSupportDir=${HELIX_CORRELATION_PAYLOAD}/build /p:EMSDK_PATH=${HELIX_CORRELATION_PAYLOAD}/build/emsdk` to the build command line.

**Comment by mdh1418** [comment]
https://github.com/dotnet/runtime/issues/49851

**Comment by radical** [comment]
The crossaot build failure is being tracked in https://github.com/dotnet/runtime/issues/50753, and is unrelated to this PR. It's breaking on `main` too.

**Comment by SamMonoRT** [comment]
@naricc - don't believe this helps in microbenchmark runs, please chime in if you see issues here.

**Comment by naricc** [comment]
@SamMonoRT 
Doing this doesn't directly help the microbenchmark runs, but we may run into similar issues with wasm-aot on the microbenchmarks that force us to use trimming on the projects BenchmarkDotnet generates, in order to have the benchmarks run in a timely manner. 

I am not exactly sure how trim-friendly BDN is. I tried using trimming for the aot-llvm runs, and it didn't work, but it was only a minor optimization in that case, so I did not spend a lot of time figuring out why.

We will probably want to do something similar for the runtime tests, but the library and runtime tests are different enough that its hard to reuse any infrastructure.

File: eng/testing/tests.wasm.targets [hunk 2]
```diff
@@ -14,49 +23,119 @@
...
     <EventSourceSupport>false</EventSourceSupport>
-    <UseSystemResourceKeys>true</UseSystemResourceKeys>
+    <UseSystemResourceKeys>false</UseSystemResourceKeys>
     <EnableUnsafeUTF7Encoding>false</EnableUnsafeUTF7Encoding>
     <HttpActivityPropagationSupport>false</HttpActivityPropagationSupport>
 
     <!-- we want to default to what Blazor has, except if we are building in Debug config -->
-    <DebuggerSupport Condition="'$(Configuration)' != 'Debug'">false</DebuggerSupport>
+    <DebuggerSupport Condition="'$(DebuggerSupport)' == '' and '$(Configuration)' != 'Debug'">false</DebuggerSupport>
   </PropertyGroup>
```
Comment (radical) [inline eng/testing/tests.wasm.targets:58]: Some these need to have values different from what `UseDefaultBlazorWasmFeatureSwitches` wants:

- `DebuggerSupport` - this is needed by some library tests because they depend on pdbs, or debugger attributes
  - Currently, I am setting this in the specific projects that need it
- `UseSystemResourceKeys` - IIUC, this is needed so that the resources don't get trimmed, which is needed by some library tests

Setting these only for some library projects makes it kinda inconsistent. And I'm wondering if that affects what the intention of `UseDefaultBlazorWasmFeatureSwitches` is.
So, do we need switch at all? If so, then what should be done here?

/cc @lewing @akoeplinger

File: eng/pipelines/common/platform-matrix.yml [hunk 3]
```diff
@@ -219,7 +219,7 @@ jobs:
       targetRid: browser-wasm
       platform: Browser_wasm
       container:
-        image: ubuntu-18.04-webassembly-20210223133559-4800846
+        image: ubuntu-18.04-webassembly-20210309143118-005aab4
```
Comment (radical) [inline eng/pipelines/common/platform-matrix.yml:222]: @steveisok @mdh1418 Do you know why this change was made? `main` is still using the older version.

**Comment by radical** [comment]
Failing test in `runtime (Libraries Test Run release mono Linux x64 Debug)`: https://github.com/dotnet/runtime/issues/51588

Failing test in `Libraries Test Run release coreclr Linux x64 Debug` is possibly: https://github.com/dotnet/runtime/issues/48658

Failing test in `Libraries Test Run release mono OSX x64 Debug`: https://github.com/dotnet/runtime/issues/51611

File: eng/pipelines/common/platform-matrix.yml (hunk 3 above)
Comment (akoeplinger) [reply eng/pipelines/common/platform-matrix.yml:222]: judging from the date it's to bring in https://github.com/dotnet/dotnet-buildtools-prereqs-docker/commit/f25f6cd55e437581a96448d8f3b66c377b95fa8a

File: eng/testing/tests.wasm.targets (hunk 2 above)
Comment (akoeplinger) [reply eng/testing/tests.wasm.targets:58]: I agree with @radical, the intention of `UseDefaultBlazorWASMFeatureSwitches` is to make us consistent with Blazor. I think overriding in the individual library test projects where necessary is fine for now.

File: eng/testing/tests.wasm.targets [hunk 4]
```diff
@@ -14,49 +23,119 @@
...
+    </ItemGroup>
+
+    <ItemGroup Condition="'$(DebuggerSupport)' == 'true'">
+      <!-- Add any pdb files, if available -->
+      <_BundlePdbFiles Include="$([System.IO.Path]::ChangeExtension('%(WasmAssembliesToBundle.Identity)', '.pdb'))" />
+      <BundleFiles Include="@(_BundlePdbFiles)" TargetDir="publish" Condition="Exists(%(_BundlePdbFiles.Identity))" />
+    </ItemGroup>
+
+    <!-- To recreate the original project on helix, we need to set the wasm properties also, same as the
+         library test project. Eg. $(InvariantGlobalization) -->
```
Comment (akoeplinger) [inline eng/testing/tests.wasm.targets:98]: do we have a way of knowing the list of properties is complete or that we need to update it?

File: src/libraries/System.Collections.NonGeneric/tests/System.Collections.NonGeneric.Tests.csproj [hunk 5]
```diff
@@ -2,6 +2,7 @@
   <PropertyGroup>
     <IncludeRemoteExecutor>true</IncludeRemoteExecutor>
     <TargetFrameworks>$(NetCoreAppCurrent)</TargetFrameworks>
+    <DebuggerSupport Condition="'$(DebuggerSupport)' == '' and '$(TargetOS)' == 'Browser'">true</DebuggerSupport>
```
Comment (akoeplinger) [inline src/libraries/System.Collections.NonGeneric/tests/System.Collections.NonGeneric.Tests.csproj:5]: curious why does this need .pdb?

File: src/tasks/WasmBuildTasks/GenerateAOTProps.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (akoeplinger) [inline src/tasks/WasmBuildTasks/GenerateAOTProps.cs:1]: this should ideally be added to https://github.com/dotnet/runtime/blob/d08f7f2074bf3c553910f4ffaf67fb008a151aef/src/tasks/tasks.proj#L8-L11

File: eng/testing/tests.wasm.targets (hunk 4 above)
Comment (radical) [reply eng/testing/tests.wasm.targets:98]: Um .. not really. So, when something breaks? I didn't add *all* the properties either, because I want to do that as it comes up.

File: src/tasks/WasmBuildTasks/GenerateAOTProps.cs (hunk 6 above)
Comment (radical) [reply src/tasks/WasmBuildTasks/GenerateAOTProps.cs:1]: I'm not sure I understand. This file will get built with `WasmBuildTasks`. What should be added to `tasks.proj`?

File: src/libraries/System.Collections.NonGeneric/tests/System.Collections.NonGeneric.Tests.csproj (hunk 5 above)
Comment (radical) [reply src/libraries/System.Collections.NonGeneric/tests/System.Collections.NonGeneric.Tests.csproj:5]: Eg:
```
  fail: [FAIL] System.Collections.Tests.HashtableTests.DebuggerAttribute
  info: System.InvalidOperationException : Expected one DebuggerDisplayAttribute on System.Collections.Hashtable.
  info:    at System.Diagnostics.DebuggerAttributes.ValidateDebuggerDisplayReferences(Object obj)
  info:    at System.Collections.Tests.HashtableTests.DebuggerAttribute()
  info:    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)

 fail: [FAIL] System.Collections.Tests.QueueTests.DebuggerAttribute_NullQueue_ThrowsArgumentNullException
  info: System.InvalidOperationException : Expected one DebuggerTypeProxyAttribute on System.Collections.Queue.
  info:    at System.Diagnostics.DebuggerAttributes.GetProxyType(Type type, Type[] genericTypeArguments)
  info:    at System.Diagnostics.DebuggerAttributes.ValidateDebuggerTypeProxyProperties(Type type, Type[] genericTypeArguments, Object obj)
  info:    at System.Diagnostics.DebuggerAttributes.ValidateDebuggerTypeProxyProperties(Type type, Object obj)
  info:    at System.Collections.Tests.QueueTests.DebuggerAttribute_NullQueue_ThrowsArgumentNullException()
  info:    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
```

**Review by lewing — APPROVED** [review]
looks green to me, there is some minor whitespace damage but this was a long road so lets land it.
